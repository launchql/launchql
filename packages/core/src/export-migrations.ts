import path from 'path';
import { mkdirSync, rmSync } from 'fs';
import { sync as glob } from 'glob';
import Case from 'case';

import { exportMeta } from './export-meta';
import { getPgPool } from 'pg-cache';
import { LaunchQLOptions } from '@launchql/types';
import { SqitchRow, writeSqitchFiles, writeSqitchPlan } from './sqitch/utils';
import { LaunchQLProject } from './class/launchql';

interface ExportMigrationsToDiskOptions {
  project: LaunchQLProject;
  options: LaunchQLOptions;
  database: string;
  databaseId: string;
  author: string;
  outdir: string;
  schema_names: string[];
  extensionName?: string;
  metaExtensionName: string;
}

interface ExportOptions {
  project: LaunchQLProject;
  options: LaunchQLOptions;
  dbInfo: {
    dbname: string;
    database_ids: string[];
  };
  author: string;
  outdir: string;
  schema_names: string[];
  extensionName?: string;
  metaExtensionName: string;
}

const exportMigrationsToDisk = async ({
  project,
  options,
  database,
  databaseId,
  author,
  outdir,
  schema_names,
  extensionName,
  metaExtensionName
}: ExportMigrationsToDiskOptions): Promise<void> => {
  outdir = outdir + '/';

  const pgPool = getPgPool({
    ...options.pg,
    database
  });

  const db = await pgPool.query(
    `select * from collections_public.database where id=$1`,
    [databaseId]
  );

  const schemas = await pgPool.query(
    `select * from collections_public.schema where database_id=$1`,
    [databaseId]
  );

  if (!db?.rows?.length) {
    console.log('NO DATABASES.');
    return;
  }

  if (!schemas?.rows?.length) {
    console.log('NO SCHEMAS.');
    return;
  }

  const name = extensionName || db.rows[0].name;

  const { replace, replacer } = makeReplacer({
    schemas: schemas.rows.filter((schema: any) =>
      schema_names.includes(schema.schema_name)
    ),
    name
  });

  const results = await pgPool.query(
    `select * from db_migrate.sql_actions order by id`
  );

  const opts = {
    name,
    replace,
    replacer,
    outdir,
    author
  };

  if (results?.rows?.length > 0) {
    await preparePackage({
      project,
      author,
      outdir,
      name,
      extensions: [
        'plpgsql',
        'uuid-ossp',
        'citext',
        'pgcrypto',
        'btree_gist',
        'postgis',
        'hstore',
        'db_meta',
        'launchql-inflection',
        'launchql-uuid',
        'launchql-utils',
        'launchql-ext-jobs',
        'launchql-jwt-claims',
        'launchql-stamps',
        'launchql-base32',
        'launchql-totp',
        'launchql-ext-types',
        'launchql-ext-default-roles'
      ]
    });

    writeSqitchPlan(results.rows, opts);
    writeSqitchFiles(results.rows, opts);

    let meta = await exportMeta({
      opts: options,
      dbname: database,
      database_id: databaseId
    });

    meta = replacer(meta);

    await preparePackage({
      project,
      author,
      outdir,
      extensions: ['plpgsql', 'db_meta', 'db_meta_modules'],
      name: metaExtensionName
    });

    const metaReplacer = makeReplacer({
      schemas: schemas.rows.filter((schema: any) =>
        schema_names.includes(schema.schema_name)
      ),
      name: metaExtensionName
    });

    const metaPackage: SqitchRow[] = [
      {
        deps: [],
        deploy: 'migrate/meta',
        content: `SET session_replication_role TO replica;
-- using replica in case we are deploying triggers to collections_public

-- unaccent, postgis affected and require grants
GRANT EXECUTE ON ALL FUNCTIONS IN SCHEMA public to public;

DO $LQLMIGRATION$
  DECLARE
  BEGIN
  
    EXECUTE format('GRANT CONNECT ON DATABASE %I TO %I', current_database(), 'app_user');
    EXECUTE format('GRANT CONNECT ON DATABASE %I TO %I', current_database(), 'app_admin');

  END;
$LQLMIGRATION$;

${meta}

UPDATE meta_public.apis
      SET dbname = current_database() WHERE TRUE;

UPDATE meta_public.sites
      SET dbname = current_database() WHERE TRUE;

SET session_replication_role TO DEFAULT;
`
      }
    ];

    opts.replacer = metaReplacer.replacer;
    opts.name = metaExtensionName;

    writeSqitchPlan(metaPackage, opts);
    writeSqitchFiles(metaPackage, opts);
  }

  pgPool.end();
};

export const exportMigrations = async ({
  project,
  options,
  dbInfo,
  author,
  outdir,
  schema_names,
  extensionName,
  metaExtensionName
}: ExportOptions): Promise<void> => {
  for (let v = 0; v < dbInfo.database_ids.length; v++) {
    const databaseId = dbInfo.database_ids[v];
    await exportMigrationsToDisk({
      project,
      options,
      extensionName,
      metaExtensionName,
      database: dbInfo.dbname,
      databaseId,
      schema_names,
      author,
      outdir
    });
  }
};


interface PreparePackageOptions {
  project: LaunchQLProject;
  author: string;
  outdir: string;
  name: string;
  extensions: string[];
}

interface Schema {
  name: string;
  schema_name: string;
}

interface MakeReplacerOptions {
  schemas: Schema[];
  name: string;
}

interface ReplacerResult {
  replacer: (str: string, n?: number) => string;
  replace: [RegExp, string][];
}

/**
 * Creates a Sqitch package directory or resets the deploy/revert/verify directories if it exists.
 */
const preparePackage = async ({
  project,
  author,
  outdir,
  name,
  extensions
}: PreparePackageOptions): Promise<void> => {
  const curDir = process.cwd();
  const sqitchDir = path.resolve(path.join(outdir, name));
  mkdirSync(sqitchDir, { recursive: true });
  process.chdir(sqitchDir);

  const plan = glob(path.join(sqitchDir, 'sqitch.plan'));
  if (!plan.length) {
    project.initModule({
      name,
      description: name,
      author,
      extensions,
    });
  } else {
    rmSync(path.resolve(sqitchDir, 'deploy'), { recursive: true, force: true });
    rmSync(path.resolve(sqitchDir, 'revert'), { recursive: true, force: true });
    rmSync(path.resolve(sqitchDir, 'verify'), { recursive: true, force: true });
  }

  process.chdir(curDir);
};

/**
 * Generates a function for replacing schema names and extension names in strings.
 */
const makeReplacer = ({ schemas, name }: MakeReplacerOptions): ReplacerResult => {
  const replacements: [string, string] = ['launchql-extension-name', name];
  const schemaReplacers: [string, string][] = schemas.map((schema) => [
    schema.schema_name,
    Case.snake(`${name}_${schema.name}`)
  ]);

  const replace: [RegExp, string][] = [...schemaReplacers, replacements].map(
    ([from, to]) => [new RegExp(from, 'g'), to]
  );

  const replacer = (str: string, n = 0): string => {
    if (!str) return '';
    if (replace[n] && replace[n].length === 2) {
      return replacer(str.replace(replace[n][0], replace[n][1]), n + 1);
    }
    return str;
  };

  return { replacer, replace };
};