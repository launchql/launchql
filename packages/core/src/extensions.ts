import { parsePlanFileSimple } from '@launchql/project-files';
import { ModuleMap } from './modules';

/**
 * Get the list of available extensions, including predefined core extensions.
 */
export const getAvailableExtensions = (
  modules: ModuleMap
): string[] => {
  const coreExtensions = [
    'address_standardizer',
    'address_standardizer_data_us',
    'bloom',
    'btree_gin',
    'btree_gist',
    'citext',
    'hstore',
    'intarray',
    'pg_trgm',
    'pgcrypto',
    'plpgsql',
    'plperl',
    'plv8',
    'postgis_tiger_geocoder',
    'postgis_topology',
    'postgis',
    'postgres_fdw',
    'unaccent',
    'uuid-ossp',
  ];

  return Object.keys(modules).reduce<string[]>((acc, module) => {
    if (!acc.includes(module)) acc.push(module);
    return acc;
  }, [...coreExtensions]);
};
