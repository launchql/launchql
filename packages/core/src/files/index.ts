// Re-export all functionality from project-files
export * from './types';
export * from './plan';
export * from './sql';
export * from './extension';
export * from './sql-scripts';