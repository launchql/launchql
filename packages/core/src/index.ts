export * from './core/class/launchql';
export * from './export/export-meta';
export * from './export/export-migrations';
export * from './extensions/extensions';
export * from './modules/modules';
export * from './packaging/package';
export * from './packaging/transform';
export * from './resolution/deps';
export * from './resolution/resolve';
export * from './workspace/paths';
export * from './workspace/utils';

// Export project-files functionality (now integrated into core)
export * from './files';
export { cleanSql } from './migrate/clean';
export { LaunchQLMigrate } from './migrate/client';
export { 
  DeployOptions, 
  DeployResult, 
  MigrateChange, 
  MigratePlanFile, 
  RevertOptions, 
  RevertResult, 
  StatusResult,
  VerifyOptions, 
  VerifyResult} from './migrate/types';
export { hashFile, hashString } from './migrate/utils/hash';
export { executeQuery,TransactionContext, TransactionOptions, withTransaction } from './migrate/utils/transaction';
