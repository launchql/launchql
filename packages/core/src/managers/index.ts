export { ConfigManager, LaunchQLConfig } from './config-manager';
export { ModuleManager, ModuleInfo } from './module-manager';
export { ExtensionManager } from './extension-manager';
export { PackageManager, PackageInfo } from './package-manager';
export { PlanGenerator } from './plan-generator';