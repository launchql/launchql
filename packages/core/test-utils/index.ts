export { TestPlan } from './TestPlan';
export { CoreDeployTestFixture } from './CoreDeployTestFixture';
export { TestFixture } from './TestFixture';

export { MigrateTestFixture } from './MigrateTestFixture';
export { TestDatabase } from './TestDatabase';
export { MigrateTestChange } from './MigrateTestChange';

export {
  FIXTURES_PATH,
  getFixturePath,
  cleanText,
  teardownAllPools,
  closeDatabasePools
} from './utils';
