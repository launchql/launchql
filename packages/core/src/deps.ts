import { readFileSync } from 'fs';
import { sync as glob } from 'glob';
import { relative, join } from 'path';
import { parsePlanFile } from './files/plan/parser';
import { ExtendedPlanFile } from './files/types';
import { LaunchQLProject } from './class/launchql';

/**
 * Represents a dependency graph where keys are module identifiers
 * and values are arrays of their direct dependencies
 */
interface DependencyGraph {
  [key: string]: string[];
}

/**
 * Result object returned by dependency resolution functions
 */
interface DependencyResult {
  /** Array of external dependencies that are not part of the current project */
  external: string[];
  /** Array of modules in topologically sorted order for deployment */
  resolved: string[];
  /** The complete dependency graph mapping modules to their dependencies */
  deps: DependencyGraph;
}

/**
 * Configuration options for customizing the core dependency resolver behavior.
 * This interface allows different dependency resolution functions to share the same
 * core algorithm while maintaining their specific behaviors.
 */
interface DependencyResolverOptions {
  /** 
   * Function to handle external dependencies when they are not found in the dependency graph.
   * Used by extDeps to add external dependencies to the external array.
   * @param dep - The dependency module name that was not found
   * @param deps - The dependency graph to potentially add the dependency to
   * @param external - Array to collect external dependencies
   */
  handleExternalDep?: (dep: string, deps: DependencyGraph, external: string[]) => void;
  
  /** 
   * Function to transform module names and resolve their dependencies during resolution.
   * Used by getDeps and resolveDependencies to handle project:module syntax and tag resolution.
   * @param module - The module name to transform
   * @param extname - The current project/extension name for context
   * @returns Object containing the transformed module name, its dependency edges, and whether to return early
   */
  transformModule?: (module: string, extname: string) => { module: string; edges: string[] | undefined; returnEarly?: boolean };
  
  /** 
   * Function to generate standardized keys for dependency lookup in the graph.
   * Defaults to the module name as-is if not provided.
   * @param module - The module name to generate a key for
   * @returns The standardized key for dependency graph lookup
   */
  makeKey?: (module: string) => string;
  
  /** 
   * The current project/extension name used for internal reference resolution.
   * This is used to distinguish between internal and external module references.
   */
  extname: string;
}

/**
 * Core dependency resolution algorithm that handles circular dependency detection
 * and topological sorting of dependencies. This unified implementation eliminates
 * code duplication between getDeps, extDeps, and resolveDependencies.
 * 
 * @param deps - The dependency graph mapping modules to their dependencies
 * @param external - Array to collect external dependencies
 * @param options - Configuration options for customizing resolver behavior
 * @returns A function that performs dependency resolution with the given configuration
 */
function createDependencyResolver(
  deps: DependencyGraph,
  external: string[],
  options: DependencyResolverOptions
) {
  const { 
    handleExternalDep, 
    transformModule, 
    makeKey = (module: string) => module,
    extname 
  } = options;

  return function dep_resolve(
    sqlmodule: string,
    resolved: string[],
    unresolved: string[]
  ): void {
    unresolved.push(sqlmodule);
    
    let moduleToResolve = sqlmodule;
    let edges: string[] | undefined;
    let returnEarly = false;

    if (transformModule) {
      const result = transformModule(sqlmodule, extname);
      moduleToResolve = result.module;
      edges = result.edges;
      returnEarly = result.returnEarly || false;
    } else {
      edges = deps[makeKey(sqlmodule)];
    }

    // Handle external dependencies if no edges found
    if (!edges) {
      if (handleExternalDep) {
        handleExternalDep(sqlmodule, deps, external);
        edges = deps[sqlmodule] || [];
      } else {
        throw new Error(`Internal module not found: ${sqlmodule}`);
      }
    }

    if (returnEarly) {
      const index = unresolved.indexOf(sqlmodule);
      unresolved.splice(index, 1);
      return;
    }

    // Process each dependency
    for (const dep of edges) {
      if (!resolved.includes(dep)) {
        if (unresolved.includes(dep)) {
          throw new Error(`Circular reference detected ${moduleToResolve}, ${dep}`);
        }
        dep_resolve(dep, resolved, unresolved);
      }
    }

    resolved.push(moduleToResolve);
    const index = unresolved.indexOf(sqlmodule);
    unresolved.splice(index, 1);
  };
}

/**
 * Generates a standardized key for SQL deployment files
 * @param sqlmodule - The module name (e.g., 'users/create')
 * @returns The standardized file path key (e.g., '/deploy/users/create.sql')
 */
const makeKey = (sqlmodule: string): string => `/deploy/${sqlmodule}.sql`;

/**
 * Resolves dependencies for SQL deployment files in a package directory.
 * This is a simplified wrapper around resolveDependencies for backward compatibility.
 * 
 * @param packageDir - The package directory containing SQL files
 * @param extname - The extension/project name for internal reference resolution
 * @returns Object containing external dependencies, resolved order, and dependency graph
 */
export const getDeps = (
  packageDir: string,
  extname: string
): DependencyResult => {
  return resolveDependencies(packageDir, extname, { tagResolution: 'preserve' });
};

/**
 * Resolves dependencies for extension modules using a pre-built module map.
 * This is a simpler version that works with module metadata rather than parsing SQL files.
 * 
 * @param name - The name of the module to resolve dependencies for
 * @param modules - Record mapping module names to their dependency requirements
 * @returns Object containing external dependencies and resolved dependency order
 */
export const extDeps = (
  name: string,
  modules: Record<string, { requires: string[] }>
): { external: string[]; resolved: string[] } => {
  if (!modules[name]) {
    throw new Error(`module ${name} does not exist!`);
  }

  const external: string[] = [];
  const deps: DependencyGraph = Object.keys(modules).reduce((memo, key) => {
    memo[key] = modules[key].requires;
    return memo;
  }, {} as DependencyGraph);

  // Handle external dependencies for extDeps - simpler than getDeps
  const handleExternalDep = (dep: string, deps: DependencyGraph, external: string[]): void => {
    external.push(dep);
    deps[dep] = [];
  };

  // Create the dependency resolver with extDeps-specific configuration
  const dep_resolve = createDependencyResolver(deps, external, {
    handleExternalDep,
    extname: name // For extDeps, we use the module name as extname
  });

  const resolved: string[] = [];
  const unresolved: string[] = [];

  dep_resolve(name, resolved, unresolved);

  return { external, resolved };
};

export interface DependencyResolutionOptions {
  /**
   * How to handle tag references:
   * - 'preserve': Keep tags as-is, as changes in plan files
   * - 'resolve': Resolve tags to their target changes
   * - 'internal': Resolve internally but preserve in output
   */
  tagResolution?: 'preserve' | 'resolve' | 'internal';
  
  /**
   * Whether to load and parse plan files for tag resolution
   * Only applicable when tagResolution is 'resolve' or 'internal'
   */
  loadPlanFiles?: boolean;
  
  /**
   * Custom plan file loader function
   * Allows overriding the default plan file loading logic
   */
  planFileLoader?: (projectName: string, currentProject: string, packageDir: string) => ExtendedPlanFile | null;
}

/**
 * Unified dependency resolution function with configurable options.
 * This is the main entry point for all dependency resolution needs,
 * supporting advanced features like tag resolution and plan file loading.
 * 
 * @param packageDir - The package directory containing SQL files
 * @param extname - The extension/project name for internal reference resolution
 * @param options - Configuration options for dependency resolution behavior
 * @returns Object containing external dependencies, resolved order, and dependency graph
 */
export const resolveDependencies = (
  packageDir: string,
  extname: string,
  options: DependencyResolutionOptions = {}
): DependencyResult => {
  const {
    tagResolution = 'preserve',
    loadPlanFiles = true,
    planFileLoader
  } = options;
  
  // For 'resolve' and 'internal' modes, we need plan file loading
  const planCache: Record<string, ExtendedPlanFile> = {};
  
  // Helper function to load a plan file for a project
  const loadPlanFile = (projectName: string): ExtendedPlanFile | null => {
    if (!loadPlanFiles) {
      return null;
    }
    
    if (planFileLoader) {
      return planFileLoader(projectName, extname, packageDir);
    }
    
    if (planCache[projectName]) {
      return planCache[projectName];
    }
    
    try {
      let planPath: string;
      if (projectName === extname) {
        // For the current project
        planPath = join(packageDir, 'launchql.plan');
      } else {
        // For external projects, use LaunchQLProject to find module path
        const project = new LaunchQLProject(packageDir);
        const moduleMap = project.getModuleMap();
        const module = moduleMap[projectName];
        
        if (!module) {
          throw new Error(`Module ${projectName} not found in workspace`);
        }
        
        const workspacePath = project.getWorkspacePath();
        if (!workspacePath) {
          throw new Error(`No workspace found for module ${projectName}`);
        }
        
        planPath = join(workspacePath, module.path, 'launchql.plan');
      }
      
      const result = parsePlanFile(planPath);
      if (result.data) {
        planCache[projectName] = result.data;
        return result.data;
      }
    } catch (error) {
      // Plan file not found or parse error
      console.warn(`Could not load plan file for project ${projectName}: ${error}`);
    }
    
    return null;
  };
  
  // Helper function to resolve a tag to the change it represents
  const resolveTagToChange = (projectName: string, tagName: string): string | null => {
    const plan = loadPlanFile(projectName);
    if (!plan) {
      return null;
    }
    
    // Find the tag
    const tag = plan.tags.find(t => t.name === tagName);
    if (!tag) {
      return null;
    }
    
    // Return the change this tag points to
    return tag.change;
  };
  
  const external: string[] = [];
  const deps: DependencyGraph = {};
  // Keep track of tag mappings for internal resolution
  const tagMappings: Record<string, string> = {};

  // Process SQL files and build dependency graph
  const files = glob(`${packageDir}/deploy/**/*.sql`);

  for (const file of files) {
    const data = readFileSync(file, 'utf-8');
    const lines = data.split('\n');
    const key = '/' + relative(packageDir, file);
    deps[key] = [];

    for (const line of lines) {
      // Handle requires statements
      const requiresMatch = line.match(/^-- requires: (.*)/);
      if (requiresMatch) {
        const dep = requiresMatch[1].trim();
        
        // For 'preserve' mode, just add the dependency as-is (like original getDeps)
        if (tagResolution === 'preserve') {
          deps[key].push(dep);
          continue;
        }
        
        // For other modes, handle tag resolution
        if (dep.includes('@')) {
          const match = dep.match(/^([^:]+):@(.+)$/);
          if (match) {
            const [, projectName, tagName] = match;
            const taggedChange = resolveTagToChange(projectName, tagName);
            
            if (taggedChange) {
              if (tagResolution === 'resolve') {
                // Full resolution: replace tag with actual change
                const resolvedDep = `${projectName}:${taggedChange}`;
                deps[key].push(resolvedDep);
              } else if (tagResolution === 'internal') {
                // Internal resolution: keep tag in deps but track mapping
                tagMappings[dep] = `${projectName}:${taggedChange}`;
                deps[key].push(dep);
              }
            } else {
              // Could not resolve tag, keep it as is
              deps[key].push(dep);
            }
          } else {
            // Invalid tag format, keep as is
            deps[key].push(dep);
          }
        } else {
          // Not a tag, keep as is
          deps[key].push(dep);
        }
        continue;
      }

      // Handle deploy statements - exactly as in original
      let m2;
      let keyToTest;

      if (/:/.test(line)) {
        m2 = line.match(/^-- Deploy ([^:]*):([\w\/]+) to pg/);
        if (m2) {
          const actualProject = m2[1];
          keyToTest = m2[2];
        
          if (extname !== actualProject) {
            throw new Error(
              `Mismatched project name in deploy file:
          Expected project: ${extname}
          Found in line   : ${actualProject}
          Line            : ${line}`
            );
          }
        
          const expectedKey = makeKey(keyToTest);
          if (key !== expectedKey) {
            throw new Error(
              `Deployment script path or internal name mismatch:
          Expected key    : ${key}
          Found in line   : ${expectedKey}
          Line            : ${line}`
            );
          }
        }

      } else {
        m2 = line.match(/^-- Deploy (.*) to pg/);
        if (m2) {
          keyToTest = m2[1];
          if (key !== makeKey(keyToTest)) {
            throw new Error(
              'deployment script in wrong place or is named wrong internally\n' + line
            );
          }
        }
      }
    }
  }

  const transformModule = (sqlmodule: string, extname: string): { module: string; edges: string[] | undefined; returnEarly?: boolean } => {
    const originalModule = sqlmodule;
    
    // For 'preserve' mode, use simpler logic (like original getDeps)
    if (tagResolution === 'preserve') {
      let moduleToResolve = sqlmodule;
      let edges = deps[makeKey(sqlmodule)];
      
      if (/:/.test(sqlmodule)) {
        // Has a prefix — could be internal or external
        const [project, localKey] = sqlmodule.split(':', 2);
        if (project === extname) {
          // Internal reference to current project
          moduleToResolve = localKey;
          edges = deps[makeKey(localKey)];
          if (!edges) {
            throw new Error(`Internal module not found: ${localKey} (from ${project}:${localKey})`);
          }
        } else {
          // External reference — always OK, even if not in deps yet
          external.push(sqlmodule);
          deps[sqlmodule] = [];
          return { module: sqlmodule, edges: [], returnEarly: true };
        }
      } else {
        // No prefix — must be internal
        if (!edges) {
          throw new Error(`Internal module not found: ${sqlmodule}`);
        }
      }
      
      return { module: moduleToResolve, edges };
    }
    
    // Check if the ORIGINAL module (before tag resolution) is external
    if (/:/.test(originalModule)) {
      const [project, localKey] = originalModule.split(':', 2);
      if (project !== extname) {
        // External reference — always OK, even if not in deps yet
        external.push(originalModule);
        deps[originalModule] = deps[originalModule] || [];
        return { module: originalModule, edges: [], returnEarly: true };
      }
    }
    
    // For internal resolution mode, check if this module is a tag and resolve it
    let moduleToResolve = sqlmodule;
    if (tagResolution === 'internal' && tagMappings[sqlmodule]) {
      moduleToResolve = tagMappings[sqlmodule];
    }
    
    let edges = deps[makeKey(moduleToResolve)];
    
    if (/:/.test(moduleToResolve)) {
      // Has a prefix — must be internal since we already handled external above
      const [project, localKey] = moduleToResolve.split(':', 2);
      if (project === extname) {
        // Internal reference to current project
        moduleToResolve = localKey;
        edges = deps[makeKey(localKey)];
        if (!edges) {
          throw new Error(`Internal module not found: ${localKey} (from ${project}:${localKey})`);
        }
      }
    } else {
      // No prefix — must be internal
      if (!edges) {
        // Check if we have edges for the original module
        edges = deps[makeKey(sqlmodule)];
        if (!edges) {
          throw new Error(`Internal module not found: ${sqlmodule}`);
        }
      }
    }
    
    // For internal resolution, process dependencies through tag mappings
    if (tagResolution === 'internal' && edges) {
      const processedEdges = edges.map(dep => {
        // Check if this dependency is external - if so, don't resolve tags
        if (/:/.test(dep)) {
          const [project, localKey] = dep.split(':', 2);
          if (project !== extname) {
            // External dependency - keep original tag name
            return dep;
          }
        }
        
        // Internal dependency - apply tag mapping if available
        if (tagMappings[dep]) {
          return tagMappings[dep];
        }
        return dep;
      });
      return { module: moduleToResolve, edges: processedEdges };
    }
    
    return { module: moduleToResolve, edges };
  };

  // Create the dependency resolver with resolveDependencies-specific configuration
  const dep_resolve = createDependencyResolver(deps, external, {
    transformModule,
    makeKey,
    extname
  });

  let resolved: string[] = [];
  const unresolved: string[] = [];

  // Add synthetic root node - exactly as in original
  deps[makeKey('apps/index')] = Object.keys(deps)
    .filter((dep) => dep.startsWith('/deploy/'))
    .map((dep) => dep.replace(/^\/deploy\//, '').replace(/\.sql$/, ''));

  dep_resolve('apps/index', resolved, unresolved);
  
  // Remove synthetic root
  const index = resolved.indexOf('apps/index');
  resolved.splice(index, 1);
  delete deps[makeKey('apps/index')];

  // Sort extensions first - exactly as in original
  const extensions = resolved.filter((module) => module.startsWith('extensions/'));
  const normalSql = resolved.filter((module) => !module.startsWith('extensions/'));
  resolved = [...extensions, ...normalSql];

  return { external, resolved, deps };
};
