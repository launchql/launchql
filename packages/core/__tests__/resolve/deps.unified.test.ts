import { TestFixture } from '../../test-utils';
import { resolveDependencies } from '../../src/deps';

let fixture: TestFixture;

beforeAll(() => {
  fixture = new TestFixture('sqitch');
});

afterAll(() => {
  fixture.cleanup();
});

describe('resolveDependencies with different options', () => {
  describe('tagResolution: preserve (default)', () => {
    it('should behave like original getDeps [simple-w-tags/2nd]', async () => {
      const res = await resolveDependencies(
        fixture.getFixturePath('simple-w-tags', 'packages', 'my-second'),
        'my-second'
      );
      expect(res).toMatchSnapshot();
    });
  });

  describe('tagResolution: resolve', () => {
    it('should resolve tags to their target changes [simple-w-tags/2nd]', async () => {
      const res = await resolveDependencies(
        fixture.getFixturePath('simple-w-tags', 'packages', 'my-second'),
        'my-second',
        { tagResolution: 'resolve' }
      );
      expect(res).toMatchSnapshot();
    });
  });

  describe('tagResolution: internal', () => {
    it('should preserve tags in output but resolve internally [simple-w-tags/2nd]', async () => {
      const res = await resolveDependencies(
        fixture.getFixturePath('simple-w-tags', 'packages', 'my-second'),
        'my-second',
        { tagResolution: 'internal' }
      );
      expect(res).toMatchSnapshot();
    });
  });

  describe('with custom plan loader', () => {
    it('should use custom plan loader when provided', async () => {
      const customLoader = jest.fn().mockReturnValue({
        project: 'my-first',
        changes: [
          { name: 'schema_myapp', note: 'Create schema' },
          { name: 'table_users', note: 'Create users table' },
          { name: 'table_products', note: 'Create products table' }
        ],
        tags: [
          { name: 'v1.0.0', change: 'table_users', note: 'First release' }
        ]
      });

      const res = await resolveDependencies(
        fixture.getFixturePath('simple-w-tags', 'packages', 'my-second'),
        'my-second',
        { 
          tagResolution: 'resolve',
          planFileLoader: customLoader
        }
      );
      
      expect(customLoader).toHaveBeenCalledWith('my-first', 'my-second', expect.any(String));
      expect(res).toMatchSnapshot();
    });
  });

  describe('with loadPlanFiles: false', () => {
    it('should not resolve tags when plan loading is disabled', async () => {
      const res = await resolveDependencies(
        fixture.getFixturePath('simple-w-tags', 'packages', 'my-second'),
        'my-second',
        { 
          tagResolution: 'resolve',
          loadPlanFiles: false
        }
      );
      expect(res).toMatchSnapshot();
    });
  });

  describe('complex scenarios', () => {
    it('should handle all modes correctly [simple-w-tags/3rd]', async () => {
      const packagePath = fixture.getFixturePath('simple-w-tags', 'packages', 'my-third');
      
      const preserveResult = await resolveDependencies(packagePath, 'my-third');
      const resolveResult = await resolveDependencies(packagePath, 'my-third', { tagResolution: 'resolve' });
      const internalResult = await resolveDependencies(packagePath, 'my-third', { tagResolution: 'internal' });
      
      expect({
        preserve: preserveResult,
        resolve: resolveResult,
        internal: internalResult
      }).toMatchSnapshot();
    });
  });
});