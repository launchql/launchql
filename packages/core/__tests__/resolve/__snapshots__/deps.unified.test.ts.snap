// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolveDependencies with different options tagResolution: internal should preserve tags in output but resolve internally [simple-w-tags/2nd] 1`] = `
{
  "deps": {
    "/deploy/create_another_table.sql": [
      "my-second:create_table",
    ],
    "/deploy/create_schema.sql": [
      "my-first:@v1.0.0",
    ],
    "/deploy/create_table.sql": [
      "my-second:create_schema",
    ],
    "my-first:@v1.0.0": [],
  },
  "external": [
    "my-first:@v1.0.0",
    "my-first:@v1.0.0",
  ],
  "resolved": [
    "create_schema",
    "create_table",
    "create_schema",
    "create_table",
    "create_another_table",
  ],
}
`;

exports[`resolveDependencies with different options tagResolution: preserve (default) should behave like original getDeps [simple-w-tags/2nd] 1`] = `
{
  "deps": {
    "/deploy/create_another_table.sql": [
      "my-second:create_table",
    ],
    "/deploy/create_schema.sql": [
      "my-first:@v1.0.0",
    ],
    "/deploy/create_table.sql": [
      "my-second:create_schema",
    ],
    "my-first:@v1.0.0": [],
  },
  "external": [
    "my-first:@v1.0.0",
    "my-first:@v1.0.0",
  ],
  "resolved": [
    "create_schema",
    "create_table",
    "create_schema",
    "create_table",
    "create_another_table",
  ],
}
`;

exports[`resolveDependencies with different options tagResolution: resolve should resolve tags to their target changes [simple-w-tags/2nd] 1`] = `
{
  "deps": {
    "/deploy/create_another_table.sql": [
      "my-second:create_table",
    ],
    "/deploy/create_schema.sql": [
      "my-first:table_users",
    ],
    "/deploy/create_table.sql": [
      "my-second:create_schema",
    ],
    "my-first:table_users": [],
  },
  "external": [
    "my-first:table_users",
    "my-first:table_users",
  ],
  "resolved": [
    "create_schema",
    "create_table",
    "create_schema",
    "create_table",
    "create_another_table",
  ],
}
`;

exports[`resolveDependencies with different options with custom plan loader should use custom plan loader when provided 1`] = `
{
  "deps": {
    "/deploy/create_another_table.sql": [
      "my-second:create_table",
    ],
    "/deploy/create_schema.sql": [
      "my-first:table_users",
    ],
    "/deploy/create_table.sql": [
      "my-second:create_schema",
    ],
    "my-first:table_users": [],
  },
  "external": [
    "my-first:table_users",
    "my-first:table_users",
  ],
  "resolved": [
    "create_schema",
    "create_table",
    "create_schema",
    "create_table",
    "create_another_table",
  ],
}
`;

exports[`resolveDependencies with different options with loadPlanFiles: false should not resolve tags when plan loading is disabled 1`] = `
{
  "deps": {
    "/deploy/create_another_table.sql": [
      "my-second:create_table",
    ],
    "/deploy/create_schema.sql": [
      "my-first:@v1.0.0",
    ],
    "/deploy/create_table.sql": [
      "my-second:create_schema",
    ],
    "my-first:@v1.0.0": [],
  },
  "external": [
    "my-first:@v1.0.0",
    "my-first:@v1.0.0",
  ],
  "resolved": [
    "create_schema",
    "create_table",
    "create_schema",
    "create_table",
    "create_another_table",
  ],
}
`;
